compiler/optimizer/op.go

```diff
@@ -140,22 +141,83 @@ func analyzeCuts(assignments []dag.Assignment, layout order.Layout) order.Layout
 	if key == nil {
 		return order.Nil
 	}
+	// This loop implements a very simple data flow analysis where we
+	// track the known order through the scoreboard.  If on exit, there
+	// is more than one field of known order, the current optimization
+	// framework cannot handle this so we return unknown (order.Nil)
+	// as a conservative stance to prevent any problematic optimizations.
+	// If there is precisely one field of known order, then that is the
+	// layout we return.  In a future version of the optimizer, we will
+	// generalize this scoreboard concept across the flowgraph for a
+	// comprehensive approach to dataflow analysis.  See issue #2756.
+	scoreboard := make(map[string]field.Path)
+	scoreboard[fieldKey(key)] = key
 	for _, a := range assignments {
-		// XXX This logic (from the original parallelize code)
-		// seems to have a bug.  See issue #2663.
-		if !fieldOf(a.RHS).Equal(key) {
-			continue
-		}
 		lhs := fieldOf(a.LHS)
+		rhs := fieldOf(a.RHS)
 		if lhs == nil {
+			// If we cannot statically determine the data flow,
+			// we give up and return unknown.  This is overly
+			// conservative in general and will miss optimization
+			// opportunities, e.g., we could do dependency
+			// analysis of a complex RHS expression.
 			return order.Nil
 		}
-		if lhs.Equal(key) {
-			return layout
+		lhsKey := fieldKey(lhs)
+		if rhs == nil {
+			// If the RHS depends on a well-defined set of fields
+			// (non of which are unambiguous like this.foo[this.bar]),
+			// and if all of such dependencies do not have an order
+			// to preserve, then we can continue along by clearing
+			// the LHS from the scoreboard knowing that is being set
+			// to something that does not have a defined order.
+			dependencies, ok := fieldsOf(a.RHS)
+			if !ok {
+				return order.Nil
+			}
+			for _, d := range dependencies {
+				key := fieldKey(d)
+				if _, ok := scoreboard[key]; ok {
+					// There's a dependency on an ordered
+					// field but we're not sophisticated
+					// enough here to know if this preserves
+					// its order...
+					return order.Nil
+				}
+			}
+			// There are no RHS dependencies on an ordered input.
+			// Clobber the LHS if present from the scoreboard and continue.
+			delete(scoreboard, lhsKey)
+			continue
+		}
+		rhsKey := fieldKey(rhs)
+		if _, ok := scoreboard[rhsKey]; ok {
+			scoreboard[lhsKey] = lhs
+			continue
+		}
+		if _, ok := scoreboard[lhsKey]; ok {
+			// LHS is in the scoreboard and the RHS isn't, so
+			// we know for sure there is no ordering guarantee
+			// on the LHS field.  So clobber it and continue.
+			delete(scoreboard, lhsKey)
 		}
-		return order.Layout{Keys: field.List{key}, Order: layout.Order}
 	}
-	return order.Nil
+	if len(scoreboard) != 1 {
+		return order.Nil
+	}
+	for _, f := range scoreboard {
+		return order.Layout{Keys: field.List{f}, Order: layout.Order}
+	}
+	panic("unreachable")
```

Comment:
Not sure this is worth the trouble.
```suggestion
	if len(scoreboard) == 0 {
		return order.Nil
	}
	var maxKey string
	for key := range scoreboard {
		if key > maxKey {
			maxKey = key
		}
	}
	return order.Layout{Keys: field.List{scoreboard[maxkey]}, Order: layout.Order}
```

Reply:
This seems more complicated to me.

Reply:
I meant that I'm not sure my suggestion is worth the trouble because it's definitely more complicated, but its incorrectness is definitely another strike against it!